### ISSUE TITLE
 different value of num query lead to output nan with detr 

### ISSUE THREAD

**Issue by daliyasia** [wang-xinyu/tensorrtx#843]
## Env

- GPU RTX2080
- OS Ubuntu18
- Cuda 11.1
- TensorRT 7

## About this repo

- branch wang-xinyu:master
- detr

## problem

I trained a new model detr from a custom dataset with 14 classes ( bg include) and set the num_querry to 50.
The model was tested on python with success. Then, I convert the model to "wts".
Finally, I changed the name query and NUM classes in the code (tensorrtx;detr cpp).
Unfortunately, the result of output[0] and output[1] are nan.
The problem is related to the num query.

Could you help me with that?

**Comment by wang-xinyu** [comment]
@freedenS can you advice?

**Comment by freedenS** [comment]
Do you have the same problem with the model in README?

**Comment by daliyasia** [comment]
Thank you for your reply   @freedenS.
No i don't have any problem with the model in README. Also, i build a new model with the same query and different classes and work very well with your code. The problem is related only on the num query. I tried to debug your code but without success

**Comment by freedenS** [comment]
Thank you for reporting!
i'll take a look.

**Comment by freedenS** [comment]
i have tried with different query numbers, but there is no problem. The result of tensorrt is same to pytorch.
can you please share your pytorch model and with it's config?
@daliyasia

**Comment by freedenS** [comment]
Btw, you'd better ensure your version of tensorrt is matching with cuda and cudnn, and cuda is up to date.

**Comment by daliyasia** [comment]
> 
Thank you for these answers.
In this [link](https://www.dropbox.com/sh/xd4ivh8av0cfqmk/AABbJIUTNrnE2wEyRc23Ivepa?dl=0) you can find the model with the following configuration : 

static const int INPUT_H = 800;
static const int INPUT_W = 800;
static const int NUM_CLASS = 14;  // include background
static const int NUM_QUERIES = 50;

**Comment by freedenS** [comment]
i have tried your model with the demo.jpg, it shows below:
![image](https://user-images.githubusercontent.com/26213470/145828712-513f633d-a3e6-470f-859c-b2da0cc9bf36.png)

**Comment by freedenS** [comment]
you can print the output of every layer with tensorrt-profile to dive into the problem

**Comment by daliyasia** [comment]
I tried to print every layer with "markOutput" and i found that, after the first conv (resnet), all value are zeroes.
Do you have an template script or link to learn how to print output layer by using tensorrt-profile ?

**Comment by freedenS** [comment]
no, i don't have. maybe you can refer to the official document.

**Comment by daliyasia** [comment]
Think you

**Comment by guods** [comment]
@daliyasia I met the same problem, the output is nan for some images.  
Have you solved the problem？

**Comment by freedenS** [comment]
@guods could you share the model with configure and images?

**Comment by guods** [comment]
@daliyasia 
static const int INPUT_H = 800;
static const int INPUT_W = 1066;
static const int NUM_CLASS = 9; // include background
static const int NUM_QUERIES =100;

**Comment by guods** [comment]
@daliyasia 
针对同一张图片，pytroch检测有目标，结果正常。TensorRT改变输出层后，在进入Tranformer之前的层一直都是有正常值的， Transformer之后进入Predict里的全连接层输出值就为nan了。问题还在Tansformer内。也不是所有的图片都有这种情况。

**Comment by daliyasia** [comment]
@guods, I am still working on it. I am trying to debug the code with markOutput to dive into the problem.
It works in python, but not on c++/tensorrt, is it the same for you?

**Comment by daliyasia** [comment]
@guods I noticed that it works for a model trained for a few epochs but as soon as I reach 100 epochs, the output is nan

**Comment by guods** [comment]
@daliyasia 我分析到在第二次Encoder中MultilAttention内的softmax层输出的值全为nan, 前一层的数据出现nan, 还未查到哪一层开始出现nan。我训练的到30几个epoch，出现这种情况，十几个epoch时没遇到。方便的话加我QQ交流吧：1120444895。

**Comment by guods** [comment]
@daliyasia Sorry, I didn't know you couldn't speak Mandarin Chinese.
" I noticed that it works for a model trained for a few epochs but as soon as I reach 100 epochs, the output is nan."
I have a similar problem, too.
With the increase of training epochs, the trained map of model is higher, but the model works even worse with TensorRT.
And before the softmax layer of the MultilAttention, the output vaule become bigger, so it is easy that the output value of the softmax layer is nan, the final output value is also nan.

**Comment by daliyasia** [comment]
@freedenS  and @wang-xinyu  could u advice me and @guods. 
@guods  and i, have the same problem.
"With the increase of training epochs, the trained map of model is higher, but the model works even worse with TensorRT.
And before the softmax layer of the MultilAttention, the output vaule become bigger, so it is easy that the output value of the softmax layer is nan, the final output value is also nan."

**Comment by freedenS** [comment]
Of course, but i need to repro this problem first.

**Comment by daliyasia** [comment]
@freedenS, to reproduce the problem mentioned before by me and @guods , you can find in this [link](https://www.dropbox.com/sh/r6iv3lfa9xw6icb/AAAEyJRPtlfNRT3OWM06RT2ea?dl=0) code, model and video test.
You have juste to add the video to a folder "data" and run ./detr -d detr100.engine /data/. Thank you in advance.

**Comment by daliyasia** [comment]
> @daliyasia Sorry, I didn't know you couldn't speak Mandarin Chinese. " I noticed that it works for a model trained for a few epochs but as soon as I reach 100 epochs, the output is nan." I have a similar problem, too. With the increase of training epochs, the trained map of model is higher, but the model works even worse with TensorRT. And before the softmax layer of the MultilAttention, the output vaule become bigger, so it is easy that the output value of the softmax layer is nan, the final output value is also nan.

I also noticed that the nan output problem is related to precision fp16. It works on FP32.

**Comment by guods** [comment]
@daliyasia Thanks for your reply. And I noticed that the nan output problem is related to precision fp16 with pytorch. It works on FP32. I try to train models with fp16.

**Comment by freedenS** [comment]
> @freedenS, to reproduce the problem mentioned before by me and @guods , you can find in this [link](https://www.dropbox.com/sh/r6iv3lfa9xw6icb/AAAEyJRPtlfNRT3OWM06RT2ea?dl=0) code, model and video test. You have juste to add the video to a folder "data" and run ./detr -d detr100.engine /data/. Thank you in advance.

sorry, this link has been deleted. can you share again?

**Comment by freedenS** [comment]
> > @daliyasia Sorry, I didn't know you couldn't speak Mandarin Chinese. " I noticed that it works for a model trained for a few epochs but as soon as I reach 100 epochs, the output is nan." I have a similar problem, too. With the increase of training epochs, the trained map of model is higher, but the model works even worse with TensorRT. And before the softmax layer of the MultilAttention, the output vaule become bigger, so it is easy that the output value of the softmax layer is nan, the final output value is also nan.
> 
> @guods I also noticed that the nan output problem is related to precision fp16. It works on FP32.

if so, you can open an issue in tensorrt
